src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
         if (ucs_likely(rdesc != NULL)) {
             rdesc->flags = 0;
             memcpy(rdesc + 1, data, length);
         } else {
-            ucs_error("ucp recv descriptor is not allocated");
-            goto out;
+            ucs_fatal("ucp recv descriptor is not allocated");
         }
     }
 
-    rdesc->length  = recv_len;
-    rdesc->hdr_len = hdr_len;
-    hdr            = data;
+    rdesc->length  = length - sizeof(ucp_stream_am_hdr_t);
+    rdesc->hdr_len = sizeof(ucp_stream_am_hdr_t);
+
+    return rdesc;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_stream_ep_enqueue(ucp_ep_ext_stream_t *ep, ucp_worker_h worker)
+{
+    ucs_list_add_tail(&worker->stream_eps, &ep->list);
+    ep->ucp_ep->flags |= UCP_EP_FLAG_STREAM_IS_QUEUED;
+}
 
-    hash_it = kh_get(ucp_worker_ep_hash, &worker->ep_hash, hdr->sender_uuid);
-    if (ucs_unlikely(hash_it == kh_end(&worker->ep_hash))) {
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_stream_am_handler(void *am_arg, void *am_data, size_t am_length,
+                      unsigned am_flags)
+{
+    ucp_worker_h        worker      = am_arg;
+    ucp_stream_am_hdr_t *hdr        = am_data;
+    ucp_ep_ext_stream_t *ep_stream  = NULL;
+    ucp_ep_h            ep;
+    ucp_recv_desc_t     *rdesc;
+    ucp_request_t       *req;
+    ssize_t             unpacked;
+
+    ucs_assert(am_length >= sizeof(ucp_stream_am_hdr_t));
+
+    rdesc = ucp_stream_am_rdesc_get(worker, am_data, am_length, am_flags);
+    ucs_assert(rdesc != NULL);
+
+    ep = ucp_worker_ep_find(worker, hdr->sender_uuid);
+    if (ucs_unlikely(ep == NULL)) {
         ucs_error("ep is not found by uuid: %lu", hdr->sender_uuid);
-        goto err;
+        goto out;
     }
 
-    ep = kh_value(&worker->ep_hash, hash_it);
+    ep_stream = ep->ext.stream;
 
-    if (!(ep->flags & UCP_EP_FLAG_STREAM_IS_QUEUED)) {
-        ep->flags |= UCP_EP_FLAG_STREAM_IS_QUEUED;
-        ucs_list_add_tail(&worker->stream_eps, &ep->ext.stream->list);
+    if (ep->flags & UCP_EP_FLAG_STREAM_IS_QUEUED) {
+        ucs_queue_push(&ep_stream->data, &rdesc->stream_queue);
+        goto out;
+    } else {
+        ucp_stream_ep_enqueue(ep_stream, worker);
+        ucs_queue_push(&ep_stream->data, &rdesc->stream_queue);
+        if (ucs_queue_is_empty(&ep_stream->reqs)) {
+            goto out;
+        }
     }
 
-    ucs_queue_push(&ep->ext.stream->data, &rdesc->stream_queue);
+    do {
+        req = ucs_queue_head_elem_non_empty(&ep_stream->reqs, ucp_request_t,
+                                            recv.queue);
+        while ((ep_stream->rdesc = ucp_stream_recv_data_nb_internal(ep_stream))) {
+
+            unpacked = ucp_stream_rdata_unpack(ucp_stream_rdesc_data(ep_stream->rdesc),
+                                               ep_stream->rdesc->length, req);
+            if (ucs_unlikely(unpacked < 0)) {
+                goto out;
+            } else if (unpacked < ep_stream->rdesc->length) {
+                ep_stream->rdesc->length -= unpacked;
```

Comment:
how does it work? looks like you unpack the beginning of the rdesc, but discard the end of it? i guess you would want to put memmove here (but pls don't)

Reply:
Need to complicate a data pattern in my tests..